---
title: "BASiCS workflow: a step-by-step analysis of expression variability using single cell RNA sequencing data"
author: 
  - name: Alan O'Callaghan
    affiliation: 
    - &MRC MRC Human Genetics Unit, Institute of Genetics \& Cancer, 
      University of Edinburgh, Western General Hospital, Crewe Road, Edinburgh, 
      EH4 2XU, UK
    email: "a.b.o'callaghan@sms.ed.ac.uk"
  - name: Nils Eling
    affiliation: 
    - &UZH Department of Quantitative Biomedicine, University of Zurich,
      Winterthurerstrasse 190, CH-8057, Zurich, Switzerland
    - &ETH Institute for Molecular Health Sciences, ETH Zurich,
      Otto-Stern Weg 7, 8093 Zurich, Switzerland
  - name: John C. Marioni
    affiliation: 
    - &EBI European Molecular Biology Laboratory, European Bioinformatics 
      Institute, Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK
    - &CRUK Cancer Research UK Cambridge Institute, University of Cambridge, 
      Li Ka Shing Centre, Cambridge, CB2 0RE, UK
  - name: Catalina A. Vallejos
    affiliation: 
    - *MRC 
    - The Alan Turing Institute, British Library, 96 Euston Road, London, 
      NW1 2DB, UK
    email: catalina.vallejos@igmm.ed.ac.uk
abstract: |
  Cell-to-cell gene expression variability is an inherent feature of complex 
  biological systems, such as immunity and development. Single-cell RNA 
  sequencing is a powerful tool to quantify this heterogeneity, but it is prone 
  to technical noise, beyond what is observed in bulk-level assays.
  In this article, we describe a step-by-step 
  computational workflow that uses the BASiCS Bioconductor package to robustly 
  quantify expression variability within and between known groups of cells (such 
  as experimental conditions or cell types). BASiCS uses an integrated framework 
  for data normalisation, technical noise quantification and downstream 
  analyses, propagating statistical uncertainty across these steps. 
  Within a single seemingly homogeneous cell population, BASiCS can identify 
  highly variable genes that exhibit strong heterogeneity as well as lowly 
  variable genes with stable expression. BASiCS also uses a probabilistic 
  decision rule to identify changes in expression variability between cell 
  populations, whilst avoiding confounding effects related to differences in 
  technical noise or in overall abundance. Using a publicly available 
  dataset, we guide users through a complete pipeline that includes 
  preliminary steps for quality control, as well as data exploration 
  using the scater and scran Bioconductor packages. The workflow is accompanied 
  by a Docker image that ensures the reproducibility of our results. 
keywords: Single-cell RNA sequencing, expression variability, 
  transcriptional noise, differential expression testing
bibliography: Workflow.bib
urlcolor: Orange
output:
  BiocWorkflowTools::f1000_article:
    fig_width: 6
    fig_height: 3.5
---

```{r setup_knitr, include = FALSE, cache = FALSE}
library("BiocStyle")
## Decide whether to display parts for BioC (TRUE) or F1000 (FALSE)
on.bioc <- FALSE
library("knitr")
library("ggplot2")
theme_set(theme_bw())
## Use fig.width = 7 for html and fig.width = 6 for pdf
## fig.width <- ifelse(on.bioc, 10, 6)
if (knitr::is_html_output()) {
  out_width <- "700px"
  out_height <- "600px"
} else if (knitr::is_latex_output()) {
  out_width <- "2.5in"
  out_height <- "3.5in"
}
knitr::opts_chunk$set(
  warning = FALSE, message = FALSE, error = FALSE,
  cache = 0, cache.path = "cache_main/",
  # fig.pos = "h",
  fig.path = "figure/"
)
options(timeout=100)
```

# Introduction

<!--- scRNA-seq and the different types of heterogeneity ---> 
<!--- Nils to revisit and add additional references if required --->
Single-cell RNA-sequencing (scRNA-seq) enables the study of genome-wide 
cell-to-cell transcriptional heterogeneity that is not captured by bulk 
experiments [@Stegle2015; @Prakadan2017; @Patange2018]. 
On the broadest level, this heterogeneity can reflect the presence of distinct 
cell subtypes or states. 
Alternatively, it can be due to gradual changes along biological processes, 
such as development and differentiation. 
Several clustering and pseudotime inference tools have been developed to
capture these types of heterogeneity [@Kiselev2019; @Saelens2019].
However, there is a limited availability of tools tailored 
to study more subtle variability within seemingly homogeneous cell populations. 
This variability can reflect deterministic or stochastic events that regulate 
gene expression and, among other settings, has been seen to increase prior to 
cell fate decisions [@Mojtahedi2016] and during ageing [@Martinez-jimenez2017]. 
Transcriptional variability has also been observed to differ from gene to gene 
and can be conserved across cell types and species [@Lin2019]. 

<!--- Narrow focus to transcriptional noise --->
Stochastic variability within a seemingly homogeneous cell population --- often 
referred to as transcriptional *noise* --- can arise from intrinsic and 
extrinsic sources [@Elowitz2002; @Eling2019]. 
Extrinsic noise refers to stochastic fluctuations induced by 
different dynamic cellular states (e.g. cell cycle, metabolism, 
intra/inter-cellular signalling) [@Zopf2013; @Iwamoto2016; @Kiviet2014]. 
In contrast, intrinsic noise arises from stochastic effects on biochemical 
processes such as transcription and translation [@Elowitz2002].
Intrinsic noise can be modulated by genetic and epigenetic modifications (such 
as mutations, histone modifications, CpG island length and nucleosome 
positioning) [@Eberwine2015; @Faure2017; @Morgan2018] and usually occurs 
at the gene level [@Elowitz2002]. 
Cell-to-cell gene expression variability estimates derived from scRNA-seq data 
capture a combination of these effects, as well as deterministic regulatory 
mechanisms [@Eling2019]. 
Moreover, these variability estimates can also be inflated by the technical 
noise that is typically observed in scRNA-seq data [@Brennecke2013]. This 
technical noise relates to systematic differences between cells that may be 
introduced by the data generating process (e.g.~due to differences in dilution 
or sequencing depth) [@Vallejos2017]. 

<!--- Experimental strategies to tackle technical noise --->
Different strategies have been incorporated into scRNA-seq protocols to control 
or attenuate technical noise. 
For example, external RNA spike-in molecules (such as the set introduced by the 
External RNA Controls Consortium, ERCC [@Rna2005]) can be added to each cellâ€™s 
lysate in a (theoretically) known fixed quantity.
Spike-ins can assist quality control steps [@McCarthy2017], data normalisation
[@Vallejos2017] and can be used to infer technical noise [@Brennecke2013].
Another strategy is to tag individual cDNA molecules using unique molecular 
identifiers (UMIs) before PCR amplification [@Islam2014]. 
Reads that contain the same UMI can be collapsed into a single molecule count,
attenuating technical variability associated to cell-to-cell differences
in amplification and sequencing depth (these technical biases are not fully 
removed unless sequencing to saturation [@Vallejos2017]). 
However, despite the benefits associated to the use of spike-ins and UMIs, 
these are not routinely available for all scRNA-seq protocols [@Haque2017]. 
In particular, spike-ins are of limited use in droplet-based protocols, 
as spike-ins can only be added to the reagent mixture in a known concentration,
and the exact quantity in each droplet necessarily remains unknown.

<!--- Introduce BASiCS --->
The Bioconductor package `r Biocpkg("BASiCS")` implements a Bayesian 
hierarchical framework that accounts for both technical and biological sources 
of noise in scRNA-seq datasets [@Vallejos2015; @Vallejos2016; @Eling2017]. The
model was initially motivated by supervised experimental designs in which 
experimental conditions correspond to groups of cells defined *a priori* (e.g. 
selected cell types obtained through FACS sorting). However, the approach can 
also be used in cases where the groups of cells of interest are computationally 
identified through clustering. In such cases, the model does not currently 
address issues associated to *post-selection inference*, where the same data is 
analysed twice: first to perform clustering and then to compare expression 
profiles between the clusters [@Lahnemann2020] (this is an inherent limitation 
of most differential expression tools). 

`r Biocpkg("BASiCS")` jointly performs data normalisation, technical noise 
quantification and downstream analyses, whilst propagating statistical 
uncertainty across these steps. 
These features are implemented within a probabilistic model that builds upon a
negative binomial distribution, a widely used statistical model in the context 
of bulk and scRNA-seq experiments [@Love2014;@Svensson2020;@Townes2020;@Townes2019]. 
The negative binomial distribution is commonly used to model count
data when the observed variability exceeds what can be captured by a simpler
Poisson model --- this is typically referred to as *over-dispersion*. 
<!--- emphasis, this is critical ---> 
Critically, `r Biocpkg("BASiCS")` enables the quantification of transcriptional
variability within a population of cells, while accounting for the overall 
mean-variance relationship that typically arises in scRNA-seq data [@Eling2018].
The latter enables the quantification of a gene-level *residual over-dispersion*
as a measure of transcriptional noise that is not confounded by differences in
gene expression. Furthermore, when available, `r Biocpkg("BASiCS")` can also 
leverage extrinsic spike-in molecules to aid data normalisation. 

<!--- Describes aims for the workflow and introduces BASiCS ---> 
This article complements existing scRNA-seq workflows based on the Bioconductor 
ecosystem (e.g. [@Lun2016; @Kim2019]), providing a detailed framework for 
transcriptional variability analyses using `r Biocpkg("BASiCS")`. 
We describe a step-by-step workflow that uses `r Biocpkg("scater")` 
[@McCarthy2017] and `r Biocpkg("scran")` [@Lun2016] to perform quality control 
(QC) as well as initial exploratory analyses.
Our analysis pipeline includes practical guidance to assess the convergence of 
the Markov Chain Monte Carlo (MCMC) algorithm that is used to infer model 
parameters in `r Biocpkg("BASiCS")`, as well as recommendations to interpret 
and post-process the model outputs. 
Finally, through a case study in the context of mouse immune cells, we illustrate
how `r Biocpkg("BASiCS")` can identify highly and lowly variable
genes within a cell population, as well as to compare expression profiles
between experimental conditions or cell types. 

All source code used to generate the results presented in this article is 
available [on Github](https://github.com/VallejosGroup/BASiCSWorkflow).
To ensure the 
reproducibility of this workflow, the analysis environment and all software 
dependencies are provided as a Docker image [@Boettiger2015]. The image 
can be obtained from
[Docker Hub](https://hub.docker.com/repository/docker/alanocallaghan/bocker).

# Methods {#methods}

This step-by-step scRNA-seq workflow is primarily based on the Bioconductor 
package ecosystem [@Amezquita2019]. 
A graphical overview is provided in Figure \@ref(fig:overview) 
and its main components are described below. 

```{r overview, out.width=out_width, out.height=out_height, fig.cap = "Graphical overview for the scRNA-seq analysis workflow described in this manuscript. Starting from a matrix of expression counts, we use the scater and scran Bioconductor packages to perform QC and initial exploratory analyses. To robustly quantify transcriptional heterogeneity within seemingly homogeneous cell populations, we apply the BASiCS Bioconductor package and  illustrate how BASiCS can be used to analyse a single or multiple pre-specified groups of cells.", echo=FALSE}
library(cowplot)
fig_svg <- cowplot::ggdraw() +
  cowplot::draw_image("figure/Overview.svg")
plot(fig_svg)
```

Here, we load all the libraries that will be used throughout this workflow.
First, we load Bioconductor libraries included in Figure \@ref(fig:overview). 
We will provide further explanation about their function in the following sections.

```{r}
library("SingleCellExperiment")
library("scater")
library("scran")
library("BASiCS")
```

Next, we load other general purpose libraries used for data visualisation (e.g.
`r CRANpkg("ggplot2")`), data manipulation (e.g. `r CRANpkg("reshape2")`) and
other libraries required to calculate some summary statistics 
(`r Biocpkg("sparseMatrixStats")`) and to assess the convergence of the 
algorithm implemented in `r Biocpkg("BASiCS")` (e.g. `r CRANpkg("coda")`). 

```{r}
# Data visualisation
library("ggplot2")
library("ggpointdensity")
library("patchwork")
library("ComplexHeatmap")
# colour scales
library("viridis")
library("circlize")
library("RColorBrewer")
# Data manipulation
library("reshape2")
library("tidyr")
# To obtain summary statistics
library("sparseMatrixStats")
# To assess convergence of the algorithm implemented in BASiCS
library("coda")
```

## Input data

We use the Bioconductor package `r Biocpkg("SingleCellExperiment")` to convert 
an input matrix of raw read counts (molecule counts for UMI-based protocols) 
into a `SingleCellExperiment` object that can also store its associated 
metadata, such as gene- and cell-specific information. 
Moreover, when available, the same object can also store read counts for 
spike-in molecules (see `help("altExp")`).
A major advantage of using `r Biocpkg("SingleCellExperiment")` is that it 
enables interoperability across a large number of Bioconductor packages [@Amezquita2019]. 

## QC and exploratory data analysis

A critical step in scRNA-seq analyses is QC, removing low quality samples that 
may distort downstream analyses. 
The [*OSCA*](https://bioconductor.org/books/release/OSCA/) online book provides 
an extensive overview on important aspects of how to perform QC of scRNA-seq 
data, including exploratory analyses [@Amezquita2019].
Here, we use QC diagnostics to identify and remove samples that correspond to 
broken cells, that are empty, or that contain multiple cells [@Ilicic2016]. 
We also remove lowly expressed genes that represent less reliable 
information.

We use the Bioconductor package `r Biocpkg("scater")` package [@McCarthy2017] to 
calculate QC metrics for each cell (e.g. total read-count) and gene 
(e.g. percentage of zeroes across all cells), respectively. 
We also use the visualisation tools implemented in the `r Biocpkg("scater")` to 
explore the input dataset and its associated QC diagnostic metrics. 
To perform further exploratory data analysis, we use the Bioconductor package 
`r Biocpkg("scran")` [@Lun2016].
The latter is used to perform *global scaling* normalisation, calculating 
cell-specific scaling factors that capture global differences in read-counts 
across cells (e.g. due to sequencing depth and PCR amplification) 
[@Lun2016pooling]. Note that, the use of `r Biocpkg("scran")` as a normalisation 
tool is not a requirement for `r Biocpkg("BASiCS")` (which uses raw read-counts 
as input). However, we recommend the use of `r Biocpkg("scran")` normalisation 
as part of the exploratory data analysis. For example, normalised read-counts 
can be used to visualise the data in a low-dimensional space (e.g. using 
principal component analysis). This can help to explore the structure of the 
data, e.g. to examine potential batch effects. 

## BASiCS - Bayesian Analysis of Single Cell Sequencing data

The `r Biocpkg("BASiCS")` Bioconductor package uses a Bayesian hierarchical 
model to simultaneously perform data normalisation, technical noise 
quantification and downstream analyses [@Vallejos2015BASiCS] [@Vallejos2016] 
[@Eling2018]. Moreover, instead of modelling expression patterns separately for 
each gene, `r Biocpkg("BASiCS")` shares information between all genes to 
robustly quantify transcriptional variability. For example, as described by 
[@Eling2018], pooling information across genes with similar mean expression 
levels helps to obtain more reliable transcriptional variability estimates. The 
latter is particularly helpful for lowly expressed genes and sparse datasets, 
where less information is available.

A high-level description of the model implemented in `r Biocpkg("BASiCS")` is
shown in Figure X (a more detailed description is provided in Section 1 of the 
Supplementary Material). Model parameters are designed to capture different 
aspects of scRNAseq data. First, *nuisance* parameters are used to (primarily) 
capture technical artifacts. This includes cell-specific parameters (indexed by 
a $j$ subscript) used to perform global scaling normalisation (similar to the 
size factors in `r Biocpkg("scran")`) and global parameters designed to capture 
technical over-dispersion (this can also be interpreted as measurement error) 
that is shared by all genes. Note that, whilst `r Biocpkg("scran")` infers 
cell-specific size factors as a pre-processing step, `r Biocpkg("BASiCS")` uses 
an integrated approach wherein data normalisation and downstream analyses are 
performed simultaneously, thereby propagating statistical uncertainty. Secondly,
`r Biocpkg("BASiCS")` summarises expression patterns within the population of 
cells under study (e.g. a experimental condition or cell type) using two sets
of gene-specific parameters (indexed by a $i$ subscript). Mean parameters 
$\mu_i$ quantify overall expression for each gene $i$. In contrast, $\delta_i$ 
captures the excess of variability that is observed with respect to what would 
be expected in a homogeneous cell population, beyond technical noise. 
The latter is used as a proxy to quantify transcriptional variability.

Three versions of the `r Biocpkg("BASiCS")` model have been published to date.
Whilst based on a similar model specification, they differ in how inference is
performed (e.g. using different priors).

and the type of downstream analysis that
can be performed. The main differences are described below (for more details, 
see Section [REF] in the Supplementary Material):

A probabilistic decision rule (based on $\delta_i$) was proposed to identify 
*highly variable genes* (HVGs) that capture the major sources of heterogeneity 
within the analysed cells [@Brennecke2013]. HVG detection is often used to 
perform feature selection, choosing the input set of genes for subsequent 
analyses. A similar rule was developed to highlight *lowly variable genes* 
(LVGs) that exhibit stable expression across the population of cells. These may 
relate to essential cellular functions and can assist the development of new 
data normalisation or integration strategies [@Lin2019].

- **Vallejos et al (2016)** [@Vallejos2016]: the model was extended to enable
**differential expression** analyses between two pre-specified groups of cells 
(e.g. different experimental conditions or cell types). This is achieved by 
comparing the posterior distribution of gene-specific parameters ($\mu_i$ and 
$\delta_i$). While several differential expression tools were previously 
proposed for scRNA-seq data (e.g. [@Kharchenko2014; @Finak2015]), some evidence 
suggests that these do not generally outperform popular bulk RNA-seq tools 
[@Soneson2018]. Moreover, most of these methods are only designed to uncover 
changes in overall expression, ignoring the more complex patterns that can arise 
at the single cell level [@Lahnemann2020]. Instead, `r Biocpkg("BASiCS")` 
embraces the high granularity of scRNA-seq data, uncovering changes in 
cell-to-cell transcriptional variability. As noted by @Vallejos2016, the inverse
relationship that is observed between mean expression and over-dispersion 
derived from (bulk and) scRNAseq can affect the interpretation of such analyses.
To address this 